3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,248 @@
+ */
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
+   * @param {!string} sentinel
+   * @param {!string} creativeId
+   */
+  constructor(parent, sentinel, creativeId) {
+    /** @private {!Window} */
+    this.parent_ = parent;
+
+    /**
+     * @private {string}
+     */
+    this.sentinel_ = sentinel;
+
+    /**
+     * @private {string}
+     */
+    this.creativeId_ = creativeId;
+
+    /**
+     * @private {string}
+     */
+    this.extraData_ = null;
+
+    /**
+     * @private {!Array<Object<string,*>>}
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * @private {!Array<!function(Array)>}
+     */
+    this.eventListeners_ = [];
+
+    if (this.parent_.getWindow().onNewAmpAnalyticsInstance) {
+      this.parent_.getWindow().onNewAmpAnalyticsInstance(this);
+    } else {
+      dev().error(TAG_, 'Vendor page must implement onNewAmpAnalyticsInstance' +
+        ' prior to loading library script.');
+    }
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur
+   * @param {!function(!Array)} listener A function that takes an array of event
+   * strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    this.eventListeners_.push(listener);
+  }
+
+  /**
+   * Receives a message from a creative for the cross-domain iframe
+   * @param Object<string,*> message The message that was received
+   */
+  receiveMessage(message) {
+    if (message[MessageTypes.ampAnalytics3pNewCreative]) {
+      this.extraData_ = message[MessageTypes.ampAnalytics3pNewCreative];
+    } else {
+      this.receivedMessagesQueue_.push(message);
+    }
+  }
+
+  /**
+   * Sends enqueued messages to the third-party vendor's metrics-collection page
+   * If there are no event listeners on the page, event messages will remain
+   * in the queue. But if there is at least one event listener, messages will be
+   * sent to those listener(s) and the queue will be emptied. This means
+   * that the first event listener will receive messages that predate its
+   * creation, but if there are additional event listeners, they may not.
+   */
+  sendQueuedMessagesToListeners() {
+    if (this.receivedMessagesQueue_.length == 0 ||
+      this.eventListeners_.length == 0) {
+      return;
+    }
+    this.eventListeners_.forEach(listener => {
+      this.dispatch_(this.receivedMessagesQueue_, listener);
+    });
+    this.receivedMessagesQueue_ = [];
+  }
+
+  /**
+   * Sends an array of messages to a single listener, using
+   * requestIdleCallback()
+   * @param {!Array<Object<string,*>>} dataToSend The array of messages
+   * @param {!function(!Array)} listener The listener
+   * @private
+   */
+  dispatch_(dataToSend, listener) {
+    this.parent_.requestIdleCallback(() => {
```

Comment:
I'm happy to change it if you want, but that assumption concerns me. I would like everything the frame does to run at a reduced priority. But since we are not explicitly setting priority==2, I think that launching all of its actions at an idle time is still a good idea.